### PR TITLE
[WIP] Fix #240 - Support XHR from live preview

### LINKED ISSUE #240: Support XHR in live dev preview via postMessage transport layer
It would be nice if users could write code like this:

```js
var xhr = new XMLHttpRequest();
xhr.open('GET', '/foo.json', true);

xhr.onload = function(e) {
  if (this.status == 200) {
    var json = this.responseText;
    ...
  }
};

xhr.send();
```

Basically, we'd be allowing the user to request files from their filesystem using XHR.  Doing this would involve the following:
* injecting a new script into the preview DOM that replaces `window.XMLHttpRequest` with a custom function
* the new `window.XMLHttpRequest` would have the same API as the original (note: there are two versions of the API, the new one is sometimes called XMLHttpRequest2, and allows for sending more types like arraybuffers, blobs, etc)
* under the hood this code would use the postMessage transport layer to send a message back to the editor side, asking for a file to be read from the filesystem.  The editor side would read the file and send it back as another `postMessage` call to the preview DOM.
* the preview DOM would get the message with the contents of the file, and call back the XHR object that's waiting for the data.

NOTE that we couldn't fully support XHR and things like synchronous XHR, since we have to do this operation async.

I've got a bit of a start done on this in nohost here https://github.com/humphd/nohost/blob/master/shims/xmlhttprequest.js.

### PR DESCRIPTION
**Pull request by gideonthomas**

### CODE REVIEW

File: src/LiveDevelopment/Agents/XHRShim.js [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        transport.send(JSON.stringify(msg));
+    }
+
+    function XMLHttpRequestLiveDev() {
+        var self = new XMLHttpRequest();
+        var requestUrl;
+        var abortCalled = false;
+
+        function decode(data) {
+            return (new global.TextDecoder('utf8')).decode(data);
```
Comment (humphd) [inline src/LiveDevelopment/Agents/XHRShim.js]: You can't rely on this being in the global, since not all browsers implement this.

Comment (gideonthomas) [inline src/LiveDevelopment/Agents/XHRShim.js]: if we're supporting text for now, I'm going to remove this method

File: src/LiveDevelopment/Agents/XHRShim.js [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+            console.log("Sending");
+
+            function handleError(error) {
+                if(typeof self.onerror === 'function') {
+                    return self.onerror(error);
+                }
+                console.error("[Brackets LiveDev] XMLHttpRequest error: ", error);
+            }
+
+            function setResponse(data) {
```
Comment (humphd) [inline src/LiveDevelopment/Agents/XHRShim.js]: Maybe throw if `data` isn't `'text'` or `'json'` for now?

**Comment by gideonthomas** [comment]
Fixed in https://github.com/humphd/brackets-browser-livedev/pull/76

**Pull request closed without merging**